tasks/synopsys-detect-task/ts/DetectScript.ts

```diff
@@ -0,0 +1,132 @@
+import {IBlackduckConfiguration} from "./model/IBlackduckConfiguration";
+import * as task from 'azure-pipelines-task-lib/task'
+import Axios, {AxiosInstance} from "axios";
+import {IProxyInfo} from "./model/IProxyInfo";
+import url from "url";
+import HttpsProxyAgent from "https-proxy-agent/dist/agent";
+import fileSystem, {WriteStream} from "fs";
+import {IDetectConfiguration} from "./model/IDetectConfiguration";
+import {parseArguments} from "./DetectUtils"
+import {IExecOptions, ToolRunner} from "azure-pipelines-task-lib/toolrunner";
+
+const fileSystemExtra = require("fs-extra")
+
+export abstract class DetectScript {
+    static readonly DETECT_DOWNLOAD_URL = "https://detect.synopsys.com"
+    scriptName: string
+
+    protected constructor(scriptName: string) {
+        this.scriptName = scriptName
+    }
+
+    abstract getTool(): ToolRunner
+
+    abstract getCommands(): Array<string>
+
+    getScriptName(): string {
+        return this.scriptName
+    }
+
+    async runScript(blackduckConfiguration: IBlackduckConfiguration, detectConfiguration: IDetectConfiguration): Promise<number> {
+        const axiosAgent: AxiosInstance = this.createAxiosAgent(blackduckConfiguration)
+        console.log("Downloading detect script.")
+        await this.downloadScript(axiosAgent, detectConfiguration.detectFolder)
+        const env = this.createEnvironmentWithVariables(blackduckConfiguration, detectConfiguration)
+        console.log("Calling detect script")
+        return await this.invokeDetect(detectConfiguration.detectFolder, env)
+    }
+
+    createAxiosAgent(blackduckData: IBlackduckConfiguration): AxiosInstance {
```

Comment:
This class seems to be doing a lot. Maybe split it out into Downloader/Runner at least.